docs/rfcs/099-safekeeper-s3-integration.md

```diff
@@ -0,0 +1,86 @@
+
+ 1. WAL backup happens synchonously, LSN is advanced each time WAL segment is uploaded
+ 2. Crash retry support, in certain cases when safekeeper restarts we may be interrupted in the middle of upload.
+    no notifications will be fired for such wal segment. Upon restart or leader re-election a catch-up operation whould enqueue complete segments beyond WAL Backup LSN 
+ 3. WAL backup is uploaded async (and potentially out of order) upon completion results are posted into a collection of ranges, WAL Backup LSN is advanced to the highest LSN for which all segments are uploaded.
+  if WAL backup LSN can not be advanced for X number of segments or Y amount of time - an Alarm is configured to bring operator attention.
+
+Each WAL upload task is an endless retry loop attempting to upload a segment; then attempt fails task is retried.
+ following are considered terminal
+ 1. a segment is uploaded - this is an succesfull completion, update stats
+ 2. Object already exists in the backup - this is success case as well, update stats.
+ 3. Leader had been re-elected - task is terminated.
+ 4. SK shutdown is requested - task is terminated.
+ 5. Segment file is not found - task is terminated, Alarm.
+ 6. All S3 Access errors are retried, Error rate is maintained, if S3 Error rate is over threshold we should alarm.
+
+
+
+## Details
+
+[![](https://mermaid.ink/img/pako:eNqllMFu2zAMhl-F8CnBHAxt0MN8KLBkzQ4t2iA69LAMgWLTqVBZ8iR5mVH03UdJsZ0ly2GYD4YhfhLJn7_8luS6wCRLLP5oUOX4RfCd4dVaAT01N07koubKwRy4hW9zXdWNw-_ncebjjJd4j1ijOQeWgVi-tFbkXAJz2vAdnnMLjy2ERGCtdfiXUmaeeKYzZjx_bepzYBVSrbDSDodEkZvD5Pb2A5WbwfPnB-B1jarYGN--dRFhEVmyDPZGONzsucxgVGubwrYpSzRjgIguAwuLjuRSjg5IBBYh7o96eu1K0LXrt1OUNpe2Vflhjco5BhnuKqSehIWykbLfF0qcZfConSjbHvMDkkhtj0qSUPEKU7COtAFpVeoP9x_jLoPB3IHZbUdXn65SuL6e0uvmZjykf1KyBa0AJZFYALuPsRn4CnwBd4qka7CvYKFNP5ie9cWuSITt0cSk1jU0ygkJtslztDau-4cmGM5n0wzIcEOATWNgFRT9SD5pDB4Jd7R54qvzDG-kGyrxqyvMtSnggT2C4apzYahzAujyIqN2XDBIbMWjJ1n-_Di2wwQO4-7NNIF5Z4BB2k44DpWwVqgd2CjhmQuFEm5jo49H4yHViQcU7sGJCqVQNP9C-OFq047_f9QkeFTl6wVV_H5RBod2XVi6kvaUBvwl_DX7FxN5skCeO_GTk7Fn3e4Lk7gwhxBL0qRCU3FR0C_vzQfWiXvBCtdJRp8FlsEryVq9E9rUBSW8KwSpmGQllxbThDdOM7qt_UKkDv_Nw-r7b0DLoqo)](https://mermaid.live/edit#pako:eNqllMFu2zAMhl-F8CnBHAxt0MN8KLBkzQ4t2iA69LAMgWLTqVBZ8iR5mVH03UdJsZ0ly2GYD4YhfhLJn7_8luS6wCRLLP5oUOX4RfCd4dVaAT01N07koubKwRy4hW9zXdWNw-_ncebjjJd4j1ijOQeWgVi-tFbkXAJz2vAdnnMLjy2ERGCtdfiXUmaeeKYzZjx_bepzYBVSrbDSDodEkZvD5Pb2A5WbwfPnB-B1jarYGN--dRFhEVmyDPZGONzsucxgVGubwrYpSzRjgIguAwuLjuRSjg5IBBYh7o96eu1K0LXrt1OUNpe2Vflhjco5BhnuKqSehIWykbLfF0qcZfConSjbHvMDkkhtj0qSUPEKU7COtAFpVeoP9x_jLoPB3IHZbUdXn65SuL6e0uvmZjykf1KyBa0AJZFYALuPsRn4CnwBd4qka7CvYKFNP5ie9cWuSITt0cSk1jU0ygkJtslztDau-4cmGM5n0wzIcEOATWNgFRT9SD5pDB4Jd7R54qvzDG-kGyrxqyvMtSnggT2C4apzYahzAujyIqN2XDBIbMWjJ1n-_Di2wwQO4-7NNIF5Z4BB2k44DpWwVqgd2CjhmQuFEm5jo49H4yHViQcU7sGJCqVQNP9C-OFq047_f9QkeFTl6wVV_H5RBod2XVi6kvaUBvwl_DX7FxN5skCeO_GTk7Fn3e4Lk7gwhxBL0qRCU3FR0C_vzQfWiXvBCtdJRp8FlsEryVq9E9rUBSW8KwSpmGQllxbThDdOM7qt_UKkDv_Nw-r7b0DLoqo)
```

Comment:
Mermaid diagrams can be included directly as a code snippet via 
"```mermaid"

Reply:
gotta check that out, the method above allows diagram rendering in VS Code though